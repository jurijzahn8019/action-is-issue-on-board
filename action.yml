name: "Is issue on a project board"
description: |
  This action will make a graphql query and determine whether an issue or pull request
  Is already added to the project board identified by the name of the board
inputs:
  token:
    description: "Github PAT with organization scope, e.g. secrets.GH_TOKEN"
    required: true
  owner:
    description: "The organization owns the board or repository"
    required: true
  board:
    description: "Project board name which should be searched for the Card with the issue"
    required: true
  number:
    description: "Issue or pull request number to look for"
    required: true
  repo:
    description: |
      the repo name (reponame in the owner scope without owner path)
      where the issue is located
    required: true
outputs:
  isOnBoard:
    description: "true if issue is added to the given board"
  boardId:
    description: node_id or graphql id of the board
  boardName: 
    description: name of teh project board
  repo:
    description: name of the repo where the issue is located
  cardId: 
    description: graphql id of the card
  columnName: 
    description: name of the column where the card with the issue is located
  columnId: 
    description: graphql id of the column where the card with the issue is located
runs:
  using: node16
  main: dist/index.js
