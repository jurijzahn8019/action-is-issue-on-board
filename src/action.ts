/* eslint-disable camelcase */
import { getInput, setFailed, setOutput, info } from "@actions/core";
import { getOctokit } from "@actions/github";
import debug from "debug";

const dbg = debug("action-is-issue-on-board:index");

export interface GqlCardResult {
  id: string;
  content?: {
    title: string;
    number: number;
    repository: {
      name: string;
      url: string;
    };
  };
}

export interface GqlColumnResult {
  name: string;
  id: string;
  cards: {
    nodes: GqlCardResult[];
  };
}

export interface GqlProjectResult {
  name: string;
  id: string;
  columns: {
    nodes: GqlColumnResult[];
  };
}

export interface GqlProjectHolderResult {
  name: string;
  projects: { nodes: GqlProjectResult[] };
}

export interface GqlResult {
  repository: GqlProjectHolderResult | null;
  organization: {
    name: string;
    projects: { nodes: GqlProjectResult[] };
  };
}

export async function run(): Promise<void> {
  dbg("Check whether issue is on the board");
  try {
    dbg("Retrieve inputs");
    const token = getInput("token", { required: true });
    const owner = getInput("owner", { required: true });
    const board = getInput("board", { required: true });
    const repo = getInput("repo", { required: true });
    const number_input = getInput("number", { required: false });
    const number = Number.parseInt(number_input, 10);

    dbg("Inputs: %s, %s, %s, %d", owner, board, repo, number);

    const query = `
    fragment RepoOfIssue on Repository { name url }
    fragment Project on Project { id name columns(first: 10) {
      nodes { id name cards(first: 100) { nodes { id content { 
              ... on Issue { title number id repository { ...RepoOfIssue } }
              ... on PullRequest { title number id repository { ...RepoOfIssue }                }
            }
          }
        }
      }}
    }
    
    query($owner: String!, $board: String!, $repo: String!) {
      repository(owner: $owner, name: $repo) { name projects(search: $board, first: 100) { nodes { ...Project } } }
      organization(login: $owner) { name projects(search: $board, first: 100) { nodes { ...Project } } }
    }    
    `;

    const client = getOctokit(token);
    const data = await client.graphql<GqlResult>({ query, owner, board, repo });

    dbg("Search On result");
    let column: GqlColumnResult | undefined;
    let card: GqlCardResult | undefined;
    let issue: GqlCardResult["content"];

    const project =
      data.repository?.projects.nodes.find((p) => p.name === board) ||
      data.organization.projects.nodes.find((p) => p.name === board);

    if (!project) {
      setOutput("isOnBoard", false);
      setFailed(`Project Board ${board} was not found`);
      return;
    }

    const found = project.columns.nodes.find((col) =>
      col.cards.nodes.find((crd) => {
        if (crd.content?.number === number) {
          column = col;
          card = crd;
          issue = crd.content;
          return true;
        }
        return false;
      })
    );

    setOutput("boardId", project.name);
    setOutput("boardName", project.id);

    if (found) {
      const iss = issue as NonNullable<GqlCardResult["content"]>;
      const crd = card as GqlCardResult;
      const col = column as GqlColumnResult;

      dbg("Issue %s is added to %s", iss.number, project.name);
      info(`Issue: ${iss.number} is on Board: ${project.name}`);

      setOutput("isOnBoard", true);
      setOutput("repo", iss.repository.name);
      setOutput("cardId", crd.id);
      setOutput("columnId", col.id);
      setOutput("columnName", col.name);
    } else {
      info(`Issue: ${number} is not on Board: ${board}`);

      setOutput("isOnBoard", false);
      setOutput("repo", undefined);
      setOutput("cardId", undefined);
      setOutput("columnId", undefined);
      setOutput("columnName", undefined);
    }
  } catch (e) {
    dbg("Failed:", e);
    setFailed((e as Error).message);
  }
}
