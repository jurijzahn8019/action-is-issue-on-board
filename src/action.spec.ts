/* eslint-disable @typescript-eslint/no-explicit-any */
import { getInput, info, setFailed, setOutput } from "@actions/core";
import { getOctokit } from "@actions/github";
import { GqlResult, run } from "./action";

jest.mock("@actions/core");
jest.mock("@actions/github");

const getInputMock = getInput as jest.MockedFunction<typeof getInput>;
const setFailedMock = setFailed as jest.MockedFunction<typeof setFailed>;
const getOctokitMock = getOctokit as jest.MockedFunction<typeof getOctokit>;
const infoMock = info as jest.MockedFunction<typeof info>;
const setOutputMock = setOutput as jest.MockedFunction<typeof setOutput>;

const octokit = {
  graphql: jest.fn() as jest.Mock<Promise<GqlResult>, any>,
};

describe("action", () => {
  beforeEach(() => {
    process.env.GITHUB_EVENT_PATH = "fixtures/review.submitted.event.json";

    getOctokitMock.mockReturnValue(octokit as any);
    getInputMock.mockReturnValueOnce("THE TOKEN");
    getInputMock.mockReturnValueOnce("the-owner");
    getInputMock.mockReturnValueOnce("the-board");
    getInputMock.mockReturnValueOnce("the-repo");
    getInputMock.mockReturnValueOnce("666");
    octokit.graphql.mockResolvedValue({
      repository: {
        name: "the-repo",
        projects: {
          nodes: [
            {
              id: "projectid",
              name: "the-board",
              columns: {
                nodes: [
                  {
                    id: "colid",
                    name: "The Column",
                    cards: {
                      nodes: [
                        {
                          id: "blabla",
                          content: {
                            title: "The Issue",
                            repository: {
                              name: "the-repo",
                              url: "https://foo.bar",
                            },
                            number: 666,
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
      organization: { name: "the-org", projects: { nodes: [] } },
    });
  });

  it("Happy Path", async () => {
    await expect(run()).resolves.toBeUndefined();

    expect(getInputMock).toHaveBeenCalledTimes(5);
    expect(getInputMock.mock.calls.map((c) => c[0])).toEqual([
      "token",
      "owner",
      "board",
      "repo",
      "number",
    ]);

    expect(getOctokitMock).toHaveBeenCalledTimes(1);
    expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN");

    expect(octokit.graphql).toHaveBeenCalledWith({
      board: "the-board",
      owner: "the-owner",
      query: `
    fragment RepoOfIssue on Repository { name url }
    fragment Project on Project { id name columns(first: 10) {
      nodes { id name cards(first: 100) { nodes { id content { 
              ... on Issue { title number id repository { ...RepoOfIssue } }
              ... on PullRequest { title number id repository { ...RepoOfIssue }                }
            }
          }
        }
      }}
    }
    
    query($owner: String!, $board: String!, $repo: String!) {
      repository(owner: $owner, name: $repo) { name projects(search: $board, first: 100) { nodes { ...Project } } }
      organization(login: $owner) { name projects(search: $board, first: 100) { nodes { ...Project } } }
    }    
    `,
      repo: "the-repo",
    });

    expect(setOutputMock).toHaveBeenCalledTimes(7);
    expect(setOutputMock.mock.calls.map((c) => c.join(": "))).toMatchSnapshot(
      "Output"
    );

    expect(infoMock).toHaveBeenCalledTimes(1);
    expect(infoMock).toHaveBeenCalledWith("Issue: 666 is on Board: the-board");

    expect(setFailedMock).not.toHaveBeenCalled();
  });

  it("Use Project from org", async () => {
    octokit.graphql.mockResolvedValue({
      repository: null,
      organization: {
        name: "the-org",
        projects: {
          nodes: [
            {
              id: "projectid",
              name: "the-board",
              columns: {
                nodes: [
                  {
                    id: "colid",
                    name: "The Column",
                    cards: {
                      nodes: [
                        {
                          id: "thenone",
                          content: undefined,
                        },
                        {
                          id: "blabla",
                          content: {
                            title: "The Issue",
                            repository: {
                              name: "the-repo",
                              url: "https://foo.bar",
                            },
                            number: 666,
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    });

    await expect(run()).resolves.toBeUndefined();

    expect(getInputMock).toHaveBeenCalledTimes(5);
    expect(getInputMock.mock.calls.map((c) => c[0])).toEqual([
      "token",
      "owner",
      "board",
      "repo",
      "number",
    ]);

    expect(getOctokitMock).toHaveBeenCalledTimes(1);
    expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN");

    expect(octokit.graphql).toHaveBeenCalledWith({
      board: "the-board",
      owner: "the-owner",
      query: `
    fragment RepoOfIssue on Repository { name url }
    fragment Project on Project { id name columns(first: 10) {
      nodes { id name cards(first: 100) { nodes { id content { 
              ... on Issue { title number id repository { ...RepoOfIssue } }
              ... on PullRequest { title number id repository { ...RepoOfIssue }                }
            }
          }
        }
      }}
    }
    
    query($owner: String!, $board: String!, $repo: String!) {
      repository(owner: $owner, name: $repo) { name projects(search: $board, first: 100) { nodes { ...Project } } }
      organization(login: $owner) { name projects(search: $board, first: 100) { nodes { ...Project } } }
    }    
    `,
      repo: "the-repo",
    });

    expect(setOutputMock).toHaveBeenCalledTimes(7);
    expect(setOutputMock.mock.calls.map((c) => c.join(": "))).toMatchSnapshot(
      "Output"
    );

    expect(infoMock).toHaveBeenCalledTimes(1);
    expect(infoMock).toHaveBeenCalledWith("Issue: 666 is on Board: the-board");

    expect(setFailedMock).not.toHaveBeenCalled();
  });

  it("Should fail on error", async () => {
    octokit.graphql.mockRejectedValue(new Error("The FOO"));

    await expect(run()).resolves.toBeUndefined();

    expect(setOutputMock).not.toHaveBeenCalled();
    expect(infoMock).not.toHaveBeenCalled();

    expect(setFailedMock).toHaveBeenCalledTimes(1);
    expect(setFailedMock).toHaveBeenCalledWith("The FOO");
  });

  it("Should fail on no project", async () => {
    octokit.graphql.mockResolvedValue({
      repository: null,
      organization: { name: "the-org", projects: { nodes: [] } },
    });

    await expect(run()).resolves.toBeUndefined();

    expect(setOutputMock).toHaveBeenCalledTimes(1);
    expect(setOutputMock).toHaveBeenCalledWith("isOnBoard", false);
    expect(infoMock).not.toHaveBeenCalled();

    expect(setFailedMock).toHaveBeenCalledTimes(1);
    expect(setFailedMock).toHaveBeenCalledWith(
      "Project Board the-board was not found"
    );
  });

  it("Should process not found", async () => {
    octokit.graphql.mockResolvedValue({
      repository: null,
      organization: {
        name: "the-org",
        projects: {
          nodes: [
            {
              id: "projectid",
              name: "the-board",
              columns: { nodes: [] },
            },
          ],
        },
      },
    });

    await expect(run()).resolves.toBeUndefined();

    expect(getInputMock).toHaveBeenCalledTimes(5);
    expect(getInputMock.mock.calls.map((c) => c[0])).toEqual([
      "token",
      "owner",
      "board",
      "repo",
      "number",
    ]);

    expect(getOctokitMock).toHaveBeenCalledTimes(1);
    expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN");

    expect(octokit.graphql).toHaveBeenCalledWith({
      board: "the-board",
      owner: "the-owner",
      query: `
    fragment RepoOfIssue on Repository { name url }
    fragment Project on Project { id name columns(first: 10) {
      nodes { id name cards(first: 100) { nodes { id content { 
              ... on Issue { title number id repository { ...RepoOfIssue } }
              ... on PullRequest { title number id repository { ...RepoOfIssue }                }
            }
          }
        }
      }}
    }
    
    query($owner: String!, $board: String!, $repo: String!) {
      repository(owner: $owner, name: $repo) { name projects(search: $board, first: 100) { nodes { ...Project } } }
      organization(login: $owner) { name projects(search: $board, first: 100) { nodes { ...Project } } }
    }    
    `,
      repo: "the-repo",
    });

    expect(setOutputMock).toHaveBeenCalledTimes(7);
    expect(setOutputMock.mock.calls.map((c) => c.join(": "))).toMatchSnapshot(
      "Output"
    );

    expect(infoMock).toHaveBeenCalledTimes(1);
    expect(infoMock).toHaveBeenCalledWith(
      "Issue: 666 is not on Board: the-board"
    );

    expect(setFailedMock).not.toHaveBeenCalled();
  });
});
