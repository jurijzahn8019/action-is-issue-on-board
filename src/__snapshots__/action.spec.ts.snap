// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action Happy Path: Output 1`] = `
Array [
  "boardId: the-board",
  "boardName: projectid",
  "isOnBoard: true",
  "repo: the-repo",
  "cardId: blabla",
  "columnId: colid",
  "columnName: The Column",
]
`;

exports[`action Should process not found: Output 1`] = `
Array [
  "boardId: the-board",
  "boardName: projectid",
  "isOnBoard: false",
  "repo: ",
  "cardId: ",
  "columnId: ",
  "columnName: ",
]
`;

exports[`action Use Project from org: Output 1`] = `
Array [
  "boardId: the-board",
  "boardName: projectid",
  "isOnBoard: true",
  "repo: the-repo",
  "cardId: blabla",
  "columnId: colid",
  "columnName: The Column",
]
`;
